src/search.py

```diff
@@ -0,0 +1,215 @@
+        return self
+
+
+class SearchResponse(BaseModel):
+    """The response model for the search endpoint."""
+
+    hits: list[HitsModel] = Field(..., description="The hits for each query sequence.")
+
+
+@lru_cache(maxsize=None)
+def _initialize_search() -> tuple[Retriever, np.ndarray]:
+    """Initialize the retriever and load the Uniprot IDs.
+
+    Returns
+    -------
+    Retriever
+        The initialized retriever.
+    np.ndarray
+        The Uniprot IDs.
+    """
+    # Load the static configuration
+    settings = get_settings()
+
+    # The encoder model always gets placed on GPU:0 relative
+    # to CUDA_VISIBLE_DEVICES. If `gpus` > 0, then the faiss index
+    # will be placed on the next available GPUs (relative to
+    # CUDA_VISIBLE_DEVICES). Otherwise, the faiss index will share
+    # the same GPU as the encoder.
+    search_gpus = 0 if settings.GPUS == 0 else list(range(1, settings.GPUS))
+
+    # Initialize the faiss index
+    faiss_index = FaissIndex(
+        dataset_dir=settings.EMBEDDING_DATASET_DIR,
+        faiss_index_path=settings.FAISS_INDEX_PATH,
+        precision=settings.SEARCH_PRECISION,
+        search_algorithm="exact",
+        search_gpus=search_gpus,
+    )
+
+    # Initialize the encoder
+    encoder = get_encoder(
+        kwargs={
+            "name": settings.ENCODER_NAME,
+            "pretrained_model_name_or_path": settings.ENCODER_PRETRAINED_MODEL_NAME_OR_PATH,
+            "dataloader_batch_size": settings.ENCODER_DATALOADER_BATCH_SIZE,
+            "dataloader_num_data_workers": settings.ENCODER_DATALOADER_NUM_DATA_WORKERS,
+        }
+    )
+
+    # Initialize the retriever
+    retriever = Retriever(faiss_index=faiss_index, encoder=encoder)
+
+    # Preload all the Uniprot IDs to avoid disk reads during the search
+    num_uniprot_ids = np.arange(len(retriever.faiss_index.dataset))
+    all_uniprot_ids = retriever.get(num_uniprot_ids, key="tags")
+
+    return retriever, all_uniprot_ids
+
+
+def search_impl(query: SearchRequest) -> SearchResponse:
+    """The search implementation."""
+    # Get the cached retriever, or initialize it if it doesn't exist
+    retriever, all_uniprot_ids = _initialize_search()
+
+    # Collect the query sequences
+    query_sequences = [x.sequence for x in query.query_sequences]
+
+    # Perform the search
+    results, query_embeddings = retriever.search(
+        query=query_sequences,
+        top_k=query.max_hits,
+        score_threshold=query.similarity_threshold,
+    )
+
+    # Convert query_embeddings to list[list[float]]
+    query_embeddings = query_embeddings.tolist()
+
+    # Loop over each query sequence and collect the hits, embeddings, and scores
+    all_hits = []
+    for indices, scores, query_embedding, query_sequence in zip(
+        results.total_indices,
+        results.total_scores,
+        query_embeddings,
+        query.query_sequences,
+    ):
+        # TODO: If there are no hits found, do we want to make best_hit None?
```

Comment:
This is the only question that came up. Right now I'm just returning an empty string for the ID and marking total_hits=0